pilot/pkg/xds/mtls_checker.go

```diff
@@ -0,0 +1,165 @@
+		subsetPolicyMode:     map[string]*networkingapi.ClientTLSSettings_TLSmode{},
+		rootPolicyMode:       mtlsModeForDefaultTrafficPolicy(dr, svcPort),
+	}
+}
+
+// mTLSDisabled returns true if the given lbEp has mTLS disabled due to any of:
+// - disabled tlsMode
+// - DestinationRule disabling mTLS on the entire host or the port
+// - PeerAuthentication disabling mTLS at any applicable level (mesh, ns, workload, port)
+func (c *mtlsChecker) isMtlsDisabled(lbEp *endpoint.LbEndpoint) bool {
+	if c == nil {
+		return false
+	}
+	_, ok := c.mtlsDisabledHosts[lbEpKey(lbEp)]
+	return ok
+}
+
+// computeForEndpoint checks destination rule, peer authentication and tls mode labels to determine if mTLS was turned off.
+func (c *mtlsChecker) computeForEndpoint(ep *model.IstioEndpoint) {
+	if drMode := c.mtlsModeForDestinationRule(ep); drMode != nil {
+		switch *drMode {
+		case networkingapi.ClientTLSSettings_DISABLE:
+			c.mtlsDisabledHosts[lbEpKey(ep.EnvoyEndpoint)] = struct{}{}
+			return
+		case networkingapi.ClientTLSSettings_ISTIO_MUTUAL:
+			// don't mark this EP disabled, even if PA or tlsMode meta mark disabled
+			return
+		}
+	}
+
+	// if endpoint has no sidecar or explicitly tls disabled by "security.istio.io/tlsMode" label.
+	if ep.TLSMode != model.IstioMutualTLSModeLabel {
+		c.mtlsDisabledHosts[lbEpKey(ep.EnvoyEndpoint)] = struct{}{}
+		return
+	}
+
+	mtlsDisabledByPeerAuthentication := func(ep *model.IstioEndpoint) bool {
+		// apply any matching peer authentications
+		peerAuthnKey := ep.Labels.String() + ":" + strconv.Itoa(int(ep.EndpointPort))
+		if value, ok := c.peerAuthDisabledMTLS[peerAuthnKey]; ok {
+			// avoid recomputing since most EPs will have the same labels/port
+			return value
+		}
+		c.peerAuthDisabledMTLS[peerAuthnKey] = factory.
+			NewMtlsChecker(c.push, ep.Namespace, ep.Labels).
+			GetMutualTLSModeForPort(ep.EndpointPort) == model.MTLSDisable
+		return c.peerAuthDisabledMTLS[peerAuthnKey]
+	}
+
+	//  mtls disabled by PeerAuthentication
+	if mtlsDisabledByPeerAuthentication(ep) {
+		c.mtlsDisabledHosts[lbEpKey(ep.EnvoyEndpoint)] = struct{}{}
+	}
+}
+
+func (c *mtlsChecker) mtlsModeForDestinationRule(ep *model.IstioEndpoint) *networkingapi.ClientTLSSettings_TLSmode {
+	if c.destinationRule == nil || len(c.destinationRule.Subsets) == 0 {
+		return c.rootPolicyMode
+	}
+
+	drSubsetKey := ep.Labels.String()
+	if value, ok := c.subsetPolicyMode[drSubsetKey]; ok {
+		// avoid recomputing since most EPs will have the same labels/port
+		return value
+	}
+
+	subsetValue := c.rootPolicyMode
+	for _, subset := range c.destinationRule.Subsets {
+		if labels.Instance(subset.Labels).SubsetOf(ep.Labels) {
+			mode := trafficPolicyTLSModeForPort(subset.TrafficPolicy, c.svcPort)
+			if mode != nil {
+				subsetValue = mode
+			}
+			break
+		}
+	}
+	c.subsetPolicyMode[drSubsetKey] = subsetValue
+	return subsetValue
+}
+
+// mtlsModeForDefaultTrafficPolicy returns true if the default traffic policy on a given dr disables mTLS
```

Comment:
Looks like incorrect description for the function

Reply:
done